kratos/spatial_containers/spatial_search_result_container.cpp

```diff
@@ -0,0 +1,481 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+// System includes
+#include <numeric>
+#include <functional>
+
+// External includes
+
+// Project includes
+#include "includes/data_communicator.h"
+#include "includes/node.h"
+#include "includes/geometrical_object.h"
+#include "spatial_containers/spatial_search_result_container.h"
+// NOTE: GlobalRetrieveGlobalPointers is an expensive function, therefore we will build the global result manually, this way avoiding the problematic coming from the need of a indexed class and the computational cost of the call. This will therefore assume that every local result doesn't have repeated values (aka not including ghost nodes in case nodes are considered)
+// #include "utilities/global_pointer_utilities.h"
+
+namespace Kratos
+{
+
+template <class TObjectType>
+SpatialSearchResultContainer<TObjectType>::SpatialSearchResultContainer()
+{
+    // TODO: Add something if required
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TObjectType>
+void SpatialSearchResultContainer<TObjectType>::AddResult(SpatialSearchResultType& rResult)
+{
+    // Check if the object has been found
+    if (rResult.IsObjectFound()) {
+        // Adding to the local results
+        auto p_result = Kratos::make_shared<SpatialSearchResultType>(rResult);
+        mLocalResults.push_back(p_result);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TObjectType>
+void SpatialSearchResultContainer<TObjectType>::AddResult(
+    TObjectType* pResult,
+    const int Rank
+    )
+{
+    // Check if the object has been found (not nullptr)
+    if (pResult != nullptr) {
+        // Adding to the local results
+        auto p_result = Kratos::make_shared<SpatialSearchResultType>(pResult, Rank);
```

Comment:
ouch, you are creating a shared pointer from a raw pointer? this is only valid is it is an intrusive pointer...is it admissible herE?

Reply:
SpatialSearchResultType does not have intrusive pointers

Reply:
I tried using intrussive pointers and it crashes

Reply:
I would need to modify GeometricalObjectBins among other classes in order to make it work, as using intrussive pointer to properly work copy constructor must be removed

Reply:
Can we accept it as it is?

Reply:
than i would say that this code is wrong...

Reply:
Can be changed to intrussive pointers, but requires refactor evething using SpatialSearchResult class, including GeometricalObjectBins, @pooyan-dadvand agrees with me

Reply:
sorry but i really do not understand how this would work. pResult is a raw pointer to an existing opject, with its own memory mangament correct?
If that is so, it is incorrect to create a shared pointer to it, as it would imply calling the destructor twice

Reply:
@pooyan-dadvand are u really sure it is admissible here?? @roigcarlo can you take a brief look?

Reply:
I explained to him this morning. I tried to use intrussive pointer, but in order to work I needed to change to many things and it would be an API breaker

Reply:
We need this ASAP, we are already far beyond all the initial DLs (this is not an excuse, but of course if I need to refactor essential code from Kratos it will take even longer). This can be fixed later (yes we will require for now more memory), as it is considered in an internal implementation of a class, the actual use of the class will not change.

Reply:
is this version "definitive" for review? i still see it depens on other PRs ...

Reply:
I think that so many changes without a scheme of what is happening is causing a lot of confusion. Let's see if we can have top down view of what is happening in the search right now ( I will make it as a code with comments )

```C++
// 1 - Creates a search structure (Bins, Tree, etc, ....)
TSearchStrcuture mSearch;

// 2 -Creates a result container, namely  "SpatialSearchResultsContainer<X>".
//       - The contents of "SearchResultsContainer<X>" are  "TSearchResults", and in particular
//          we are using "GeometricalObject" as X in most of the new tests.
SpatialSearchResultsContainer<TSearchStrcuture::ResultType> searchResults 

// 3- Performs the search and adds the result to the result container
auto result = TSearchStructure.SearchInRadius(....)
result.whatever_post_operation();

searchResults.add(result)
```

As I see here, the problem is in the `SearchResultType`, not in the `shared_ptr<SearchStructureContainerType>`. 

I understand this requires to refactor every `SearchResultType` so its intrusive, but this a new addition and I think is legit to ask for such things. Currently there is no `SearchResultType`in Kratos that has that problem as we are using `ResultType` in all the search structures directly either from the type of results of something custom that the user has created while instantiating the SearchStructure.

For example:
```C++
  typedef TConfigure                                    Configure;
  typedef typename TConfigure::PointType                PointType;
  typedef typename TConfigure::PointerType              PointerType;
  typedef typename TConfigure::ContainerType            ContainerType;
  typedef typename TConfigure::IteratorType             IteratorType;
  typedef typename TConfigure::ResultContainerType      ResultContainerType;
  typedef typename TConfigure::ResultIteratorType       ResultIteratorType;
  typedef typename TConfigure::DistanceIteratorType     DistanceIteratorType;
```

and the default node configure class is defined as:
```c++
    typedef Point                            PointType;
    typedef Node                           ObjectType;
    typedef ObjectType::Pointer     PointerType;

    typedef PointerVectorSet<ObjectType,IndexedObject>  ObjectContainerType;
```

So Results are intrusive by design.

What I think @RiccardoRossi is complaining about is that since `SearchResultType` is not intrusive but raw, you may destroy de memory at any moment leaving the `shared_ptr<SearchStructureContainerType>` as a valid shared_ptr pointing to a container full of invalid references.

Personally I agree with this observation particularly because as I understand the code (maybe I am wrong), right now the default behavior is to return directly the result type from the `PointType` aka `Element`, `Condition`, `Node`
or give freedom to the user, but with the addition of the `SearchResultType` **we are enforcing a non intrusive_ptr** as the way to go for everyone.

Reply:
@loumalouomega  this is NOT resolved.

I am NOT telling that this must be an intrusive pointer. I am telling that if you already created an object, you cannot later take ownership of it by a shared pointer.

I am going to post something in the code a few lines above this, since my concern may be clearer there than here.

Reply:
> is this version "definitive" for review? i still see it depens on other PRs ...

This is the definitive version, except for the pending PR, but you can already grasp the final version.

Reply:
> I think that so many changes without a scheme of what is happening is causing a lot of confusion. Let's see if we can have top down view of what is happening in the search right now ( I will make it as a code with comments )
> 
> ```c++
> // 1 - Creates a search structure (Bins, Tree, etc, ....)
> TSearchStrcuture mSearch;
> 
> // 2 -Creates a result container, namely  "SpatialSearchResultsContainer<X>".
> //       - The contents of "SearchResultsContainer<X>" are  "TSearchResults", and in particular
> //          we are using "GeometricalObject" as X in most of the new tests.
> SpatialSearchResultsContainer<TSearchStrcuture::ResultType> searchResults 
> 
> // 3- Performs the search and adds the result to the result container
> auto result = TSearchStructure.SearchInRadius(....)
> result.whatever_post_operation();
> 
> searchResults.add(result)
> ```
> 
> As I see here, the problem is in the `SearchResultType`, not in the `shared_ptr<SearchStructureContainerType>`.
> 
> I understand this requires to refactor every `SearchResultType` so its intrusive, but this a new addition and I think is legit to ask for such things. Currently there is no `SearchResultType`in Kratos that has that problem as we are using `ResultType` in all the search structures directly either from the type of results of something custom that the user has created while instantiating the SearchStructure.
> 
> For example:
> 
> ```c++
>   typedef TConfigure                                    Configure;
>   typedef typename TConfigure::PointType                PointType;
>   typedef typename TConfigure::PointerType              PointerType;
>   typedef typename TConfigure::ContainerType            ContainerType;
>   typedef typename TConfigure::IteratorType             IteratorType;
>   typedef typename TConfigure::ResultContainerType      ResultContainerType;
>   typedef typename TConfigure::ResultIteratorType       ResultIteratorType;
>   typedef typename TConfigure::DistanceIteratorType     DistanceIteratorType;
> ```
> 
> and the default node configure class is defined as:
> 
> ```c++
>     typedef Point                            PointType;
>     typedef Node                           ObjectType;
>     typedef ObjectType::Pointer     PointerType;
> 
>     typedef PointerVectorSet<ObjectType,IndexedObject>  ObjectContainerType;
> ```
> 
> So Results are intrusive by design.
> 
> What I think @RiccardoRossi is complaining about is that since `SearchResultType` is not intrusive but raw, you may destroy de memory at any moment leaving the `shared_ptr<SearchStructureContainerType>` as a valid shared_ptr pointing to a container full of invalid references.
> 
> Personally I agree with this observation particularly because as I understand the code (maybe I am wrong), right now the default behavior is to return directly the result type from the `PointType` aka `Element`, `Condition`, `Node` or give freedom to the user, but with the addition of the `SearchResultType` **we are enforcing a non intrusive_ptr** as the way to go for everyone.

Do you have a moment to discuss, because I think we are mixing things